src/lib.rs

```diff
@@ -379,19 +386,27 @@ mod test {
             .address_to_nonce
             .insert(contract_address, nonce);
 
-        let mut state = CachedState::new(Arc::new(state_reader), HashMap::new());
+        let state = CachedState::new(
+            Arc::new(state_reader),
+            Arc::new(PermanentContractClassCache::default()),
+        );
 
         // Initialize state.contract_classes
-        let contract_classes = HashMap::from([(
+        state.contract_class_cache().set_contract_class(
             class_hash,
             CompiledClass::Deprecated(Arc::new(contract_class)),
-        )]);
-        state.set_contract_classes(contract_classes).unwrap();
+        );
 
         let mut block_context = BlockContext::default();
         block_context.starknet_os_config.gas_price = 1;
 
-        let estimated_fee = estimate_message_fee(&l1_handler, state, &block_context).unwrap();
+        let estimated_fee = estimate_message_fee(
+            &l1_handler,
+            state.clone(),
+            state.contract_class_cache().clone(),
```

Comment:
Isn't the new argument redundant? `state` already has its own copy of the cache at that point.

Reply:
It may seem redundant, but how would you extract the contract cache from a `StateReader` otherwise? The contract class cache is necessary to create the `CachedState`, but the `StateReader` doesn't provide us with a way to retrieve it.

We could add a method to the `StateReader` trait which returns the cache but I'm not sure the `StateReader` is required to have a contract class cache.

Reply:
Ah wait, maybe we could send in the `CachedState` directly instead of having a `CachedState<CachedState<InMemoryStateReader, PermanentContractCache>, PermanentContractCache>`. Let me try if that'd work.